### PR TITLE
Add baseline version test

### PR DESCRIPTION
**Pull request by mikaelm12**
Issue: https://github.com/aspnet/AspNetCore/issues/7546

### CODE REVIEW

**Review by natemcmaster — APPROVED** [review]
:shipit: once you have approval from @Eilon or @muratg, too.

This change is a test change only and gives us coverage which would have caught #7546

File: test/SharedFx.UnitTests/SharedFxTests.cs [hunk 1]
```diff
@@ -1,14 +1,76 @@
...
+        {
+            var previousVersion = TestData.GetPreviousAspNetCoreReleaseVersion();
+            var url = $"https://dotnetcli.blob.core.windows.net/dotnet/aspnetcore/Runtime/" + previousVersion + "/aspnetcore-runtime-internal-" + previousVersion + "-win-x64.zip";
+            var zipName = "assemblies.zip";
+            var nugetAssemblyVersions = new Dictionary<string, Version>();
+
+            var root = TestData.GetDotNetRoot();
+            var dir = Path.Combine(root, "shared", config.Name, config.Version);
+
+            using (var testClient = new WebClient())
```
Comment (Eilon) [inline test/SharedFx.UnitTests/SharedFxTests.cs]: This definitely needs retry logic. Blob storage can be very finicky.

Here's an amazing RetryHelper: https://github.com/aspnet/AspNetCore-Internal/blob/master/Infrastructure/src/TriageBuildFailures/RetryHelpers.cs

And here's an example of how it's used:
https://github.com/aspnet/AspNetCore-Internal/blob/018295b06b95ff24e016cb34151785fd30eb06b7/Infrastructure/src/TriageBuildFailures/GitHub/GitHubClientWrapper.cs#L58

File: test/SharedFx.UnitTests/SharedFxTests.cs [hunk 2]
```diff
@@ -1,14 +1,76 @@
...
+            {
+                testClient.DownloadFile(url, zipName);
+            }
+
+            var zipPath = Path.Combine(AppContext.BaseDirectory, zipName);
+
+            if (!Directory.Exists("unzipped"))
```
Comment (Eilon) [inline test/SharedFx.UnitTests/SharedFxTests.cs]: Should this create a new, random folder somewhere? What if there's already junk there from a previous failed run?

Comment (Eilon) [reply test/SharedFx.UnitTests/SharedFxTests.cs]: Also, where is the "current directory" when this is called? Looks like "current directory" and "BaseDirectory" are used but are they necessarily the same?

File: test/SharedFx.UnitTests/SharedFxTests.cs [hunk 3]
```diff
@@ -1,14 +1,76 @@
...
+            {
+                ZipFile.ExtractToDirectory(zipPath, "unzipped");
+            }
+
+            var nugetAssembliesPath = Path.Combine(AppContext.BaseDirectory, "unzipped", "shared", config.Name, previousVersion);
+
+            string[] files = Directory.GetFiles(nugetAssembliesPath, "*.dll");
```
Comment (Eilon) [inline test/SharedFx.UnitTests/SharedFxTests.cs]: `var`

File: test/SharedFx.UnitTests/SharedFxTests.cs [hunk 4]
```diff
@@ -1,14 +1,76 @@
...
+            foreach (string file in files)
```
Comment (Eilon) [inline test/SharedFx.UnitTests/SharedFxTests.cs]: More `var`

File: test/SharedFx.UnitTests/SharedFxTests.cs [hunk 5]
```diff
@@ -1,14 +1,76 @@
...
+            {
+                try
+                {
+                    var assemblyVersion = AssemblyName.GetAssemblyName(file)?.Version;
+                    var splitPath = file.Split('\\');
+                    var dllName = splitPath[splitPath.Length - 1];
```
Comment (Eilon) [inline test/SharedFx.UnitTests/SharedFxTests.cs]: Is this trying to do this? https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfilename?view=netframework-4.7.2

File: test/SharedFx.UnitTests/SharedFxTests.cs [hunk 6]
```diff
@@ -1,14 +1,76 @@
...
+            }
+
+            var nugetAssembliesPath = Path.Combine(AppContext.BaseDirectory, "unzipped", "shared", config.Name, previousVersion);
+
+            string[] files = Directory.GetFiles(nugetAssembliesPath, "*.dll");
+            foreach (string file in files)
+            {
+                try
+                {
+                    var assemblyVersion = AssemblyName.GetAssemblyName(file)?.Version;
```
Comment (Eilon) [inline test/SharedFx.UnitTests/SharedFxTests.cs]: Is the `?` needed here? If so, why?

**Review by Eilon — CHANGES_REQUESTED** [review]

File: test/SharedFx.UnitTests/SharedFxTests.cs [hunk 7]
```diff
@@ -1,14 +1,76 @@
...
+            files = Directory.GetFiles(dir, "*.dll");
+
+            Assert.All(files, file =>
+            {
+                try
+                {
+                    var localAssemblyVersion = AssemblyName.GetAssemblyName(file)?.Version;
+                    var splitPath = file.Split('\\');
+                    var dllName = splitPath[splitPath.Length - 1];
+                    Assert.True(nugetAssemblyVersions.ContainsKey(dllName), $"Expected {dllName} to be in the downloaded dlls");
```
Comment (Eilon) [inline test/SharedFx.UnitTests/SharedFxTests.cs]: Oh and BTW just checking, are there are no xUnit asserts for these two asserts here? Usually `Assert.True` is frowned upon, unless there's no other reasonable choice.

**Review by Eilon — APPROVED** [review]
Looks good. Please send TELL MODE email to shiproom alias with FYI (specify it's for 2.1). I sent you an email with info.

File: test/SharedFx.UnitTests/SharedFx.UnitTests.csproj [hunk 8]
```diff
@@ -34,6 +38,7 @@
     <PackageReference Include="xunit" Version="$(XunitPackageVersion)" />
     <PackageReference Include="xunit.analyzers" Version="$(XunitAnalyzersPackageVersion)" />
     <PackageReference Include="xunit.runner.visualstudio" Version="$(XunitRunnerVisualstudioPackageVersion)" />
+    <PackageReference Include="McMaster.Extensions.CommandLineUtils" Version="2.2.2" />
```
Comment (dougbu) [inline test/SharedFx.UnitTests/SharedFx.UnitTests.csproj]: I know this is test-only but have we used @natemcmaster's private packages elsewhere?

File: test/SharedFx.UnitTests/SharedFxTests.cs (hunk 7 above)
Comment (dougbu) [reply test/SharedFx.UnitTests/SharedFxTests.cs]: `Assert.Contains(nugetAssemblyVersions.Keys, ...)` should work here. The only supposed loss would be the custom error message when it fails.

**Review by dougbu — APPROVED** [review]

File: test/SharedFx.UnitTests/SharedFx.UnitTests.csproj (hunk 8 above)
Comment (mikaelm12) [reply test/SharedFx.UnitTests/SharedFx.UnitTests.csproj]: I mimicked here what we did in the internal repo.

File: test/SharedFx.UnitTests/SharedFxTests.cs (hunk 7 above)
Comment (mikaelm12) [reply test/SharedFx.UnitTests/SharedFxTests.cs]: We could change to use the `Assert.Contains` api. The error message becomes slightly less clear. I'm indifferent.

**Pull request merged**